dsew_community_profile/delphi_dsew_community_profile/pull.py

```diff
@@ -0,0 +1,273 @@
+            self.dfs[(sheet.level, sig)] = pd.concat([
+                pd.DataFrame({
+                    "geo_id": sheet.geo_id_select(df).apply(sheet.geo_id_apply),
+                    "timestamp": pd.to_datetime(self.times[si[0]][sig]),
+                    "val": df[si[-2]],
+                    "se": None,
+                    "sample_size": None
+                })
+                for si in sig_select
+            ])
+        self.dfs[(sheet.level, "total")]["val"] /= 7 # 7-day total -> 7-day average
+
+
+def as_cached_filename(params, config):
+    """Formulate a filename to uniquely identify this report in the input cache."""
+    return os.path.join(
+        params['indicator']['input_cache'],
+        f"{config['assetId']}--{config['filename']}"
+    )
+
+def fetch_listing(params):
+    """Generate the list of report files to process."""
+    listing = requests.get(DOWNLOAD_LISTING).json()['metadata']['attachments']
+
+    # drop the pdf files
+    listing = [
+        dict(el, cached_filename=as_cached_filename(params, el))
+        for el in listing if el['filename'].endswith("xlsx")
+    ]
+
+    if params['indicator']['reports'] == 'new':
+        # drop files we already have in the input cache
+        listing = [el for el in listing if not os.path.exists(el['cached_filename'])]
+    elif params['indicator']['reports'].find("--") > 0:
+        # drop files outside the specified publish-date range
+        start_str, _, end_str = params['indicator']['reports'].partition("--")
+        start_date = datetime.datetime.strptime(start_str, "%Y-%m-%d").date()
+        end_date = datetime.datetime.strptime(end_str, "%Y-%m-%d").date()
+        def keep(attachment):
+            publish_date = Dataset.parse_publish_date(attachment['filename'])
+            return start_date <= publish_date <= end_date
+        listing = [el for el in listing if keep(el)]
+    return listing
+
+def download_and_parse(listing, logger):
+    """Convert a list of report files into Dataset instances."""
+    datasets = {}
+    for item in listing:
+        d = Dataset(item, logger=logger)
+        for sig, df in d.dfs.items():
+            if sig not in datasets:
+                datasets[sig] = []
+            datasets[sig].append(df)
+    return datasets
+
+def fetch_new_reports(params, logger=None):
+    """Retrieve, compute, and collate all data we haven't seen yet."""
+    listing = fetch_listing(params)
+
+    # download and parse individual reports
+    datasets = download_and_parse(listing, logger)
+
+    # collect like signals together
+    ret = {}
+    for sig, lst in datasets.items():
+        ret[sig] = pd.concat(lst)
+
+    # add nation from state
+    geomapper = GeoMapper()
+    for sig in SIGNALS:
+        df = geomapper.replace_geocode(
+            ret[("state", sig)].rename(columns={"geo_id":"state_code"}),
+            'state_code',
+            'nation',
+            new_col="geo_id"
+        )
```

Comment:
This sums values across all states, which is valid for the "total" signal but gives the wrong value for "positivity". Since it's a rate, "positivity" should be averaged across states instead.

I was comparing output for Nov 13 nation positivity (4.05) vs the value given in the official report (0.065) for Nov 16. An unweighted mean of the 7-day avg values gets pretty close (0.078).

Our output for Nov 9 nation "total" matches the official reported value for Nov 16 to three decimal places after converting to 7-day avg.

Reply:
Fixed in ac57ecc

Reply:
hmm apparently I completely hallucinated that fix as it is not, in fact, there. let me poke a bit

Reply:
really actually fixed in e363c06